File: src/MainLib/Marqdouj.DotNet.AzureMaps.Blazor/Marqdouj.DotNet.AzureMaps.Blazor/Models/Interop/Modules/AzConfigurations.cs
using Marqdouj.DotNet.AzureMaps.Blazor.Models.Configuration;
using Marqdouj.DotNet.AzureMaps.Blazor.Models.GeoJson;
using Microsoft.JSInterop;
using System.Runtime.CompilerServices;

namespace Marqdouj.DotNet.AzureMaps.Blazor.Models.Interop.Modules
{
    public interface IAzMapConfigurations
    {
        Task<CameraOptionsGet> GetCamera();
        Task<MapOptionsGet> GetMapOptions();
        Task<ServiceOptions> GetServiceOptions();
        Task<StyleOptions> GetStyle();
        Task<TrafficOptions> GetTraffic();
        Task<UserInteractionOptions> GetUserInteraction();
        Task SetCamera(CameraOptions? camera, CameraBoundsOptionsSet? cameraBounds = null, AnimationOptions? animation = null);
        Task SetMapOptions(MapOptionsSet options);
        Task SetServiceOptions(ServiceOptions options);
        Task SetStyle(StyleOptions options);
        Task SetTraffic(TrafficOptions? options);
        Task SetUserInteraction(UserInteractionOptions options);
        Task ZoomTo(Position center, double zoomLevel);
    }

    internal class AzConfigurations(IAzMapModuleReferences dotNetRef, string mapId) : IAzMapConfigurations
    {
        private readonly IAzMapModuleReferences dotNetRef = dotNetRef;
        private readonly string mapId = mapId;
        private IJSRuntime JsRuntime => dotNetRef.JsRuntime;

        #region Camera
        public async Task<CameraOptionsGet> GetCamera()
        {
            return await JsRuntime.InvokeAsync<CameraOptionsGet>(GetJsInteropMethod(), mapId);
        }

        public async Task SetCamera(CameraOptions? camera, CameraBoundsOptionsSet? cameraBounds = null, AnimationOptions? animation = null)
        {
            await JsRuntime.InvokeVoidAsync(GetJsInteropMethod(), mapId, new CameraOptionsSet(camera, cameraBounds, animation));
        }

        public async Task ZoomTo(Position center, double zoomLevel)
        {
            var camera = await GetCamera();
            var options = camera.ToCameraOptions();
            options.Center = center;
            options.Zoom = zoomLevel;
            await SetCamera(options);
        }
        #endregion

        #region MapOptions
        public async Task SetMapOptions(MapOptionsSet options)
        {
            await JsRuntime.InvokeVoidAsync(GetJsInteropMethod(), mapId, options);
        }

        public async Task<MapOptionsGet> GetMapOptions()
        {
            return await JsRuntime.InvokeAsync<MapOptionsGet>(GetJsInteropMethod(), mapId);
        }
        #endregion

        #region Service
        public async Task<ServiceOptions> GetServiceOptions()
        {
            return await JsRuntime.InvokeAsync<ServiceOptions>(GetJsInteropMethod(), mapId);
        }

        public async Task SetServiceOptions(ServiceOptions options)
        {
            await JsRuntime.InvokeVoidAsync(GetJsInteropMethod(), mapId, options);
        }
        #endregion

        #region Style
        public async Task<StyleOptions> GetStyle()
        {
            return await JsRuntime.InvokeAsync<StyleOptions>(GetJsInteropMethod(), mapId);
        }

        public async Task SetStyle(StyleOptions options)
        {
            await JsRuntime.InvokeVoidAsync(GetJsInteropMethod(), mapId, options);
        }
        #endregion

        #region Traffic
        public async Task<TrafficOptions> GetTraffic()
        {
            return await JsRuntime.InvokeAsync<TrafficOptions>(GetJsInteropMethod(), mapId);
        }

        public async Task SetTraffic(TrafficOptions? options)
        {
            await JsRuntime.InvokeVoidAsync(GetJsInteropMethod(), mapId, options);
        }
        #endregion

        #region UserInteraction
        public async Task<UserInteractionOptions> GetUserInteraction()
        {
            return await JsRuntime.InvokeAsync<UserInteractionOptions>(GetJsInteropMethod(), mapId);
        }

        public async Task SetUserInteraction(UserInteractionOptions options)
        {
            await JsRuntime.InvokeVoidAsync(GetJsInteropMethod(), mapId, options);
        }
        #endregion

        private static string GetJsInteropMethod([CallerMemberName] string name = "")
            => ModuleExtensions.GetJsModuleMethod(JsModule.Configurations, name);
    }
}

File: src/MainLib/Marqdouj.DotNet.AzureMaps.Blazor/Marqdouj.DotNet.AzureMaps.Blazor/scripts/azmaps/modules/AzConfigurations.ts
import * as atlas from "azure-maps-control"
import { AzFactory } from "./AzFactory";

export class AzConfigurations {
    // #region Camera
    public static getCamera(mapId: string): any {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        return map.getCamera();
    }

    public static setCamera(mapId: string, options: CameraOptionsSet): void {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        let cameraOptions: TCameraOptionsSet = AzConfigurations.#buildCameraOptionsSet(options);
        map.setCamera(cameraOptions);
    }

    static #buildCameraOptionsSet(options: CameraOptionsSet): TCameraOptionsSet {
        let cameraOptions: TCameraOptionsSet = {};

        if (options.camera) {
            cameraOptions = { ...options.camera, ...options.animation };
        }
        else if (options.cameraBounds) {
            cameraOptions = { ...options.cameraBounds, ...options.animation };
        }

        return cameraOptions;
    }

    // #endregion

    // #region MapOptions
    public static getMapOptions(mapId: string): any {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        const mapCamera = map.getCamera();
        const service = map.getServiceOptions();
        const style = map.getStyle();
        const userInteraction = map.getUserInteraction();

        return {
            mapCamera: mapCamera,
            service: service,
            style: style,
            userInteraction: userInteraction
        };
    }

    public static setMapOptions(mapId: string, options: MapOptionsSet): void {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        const cameraOptions: TCameraOptionsSet = AzConfigurations.#buildCameraOptionsSet(options.camera);
        map.setCamera(cameraOptions);

        if (options.service) {
            map.setServiceOptions(options.service);
        }
        if (options.style) {
            map.setStyle(options.style);
        }
        if (options.userInteraction) {
            map.setUserInteraction(options.userInteraction);
        }
    }
    // #endregion

    // #region Service
    public static getServiceOptions(mapId: string): any {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        return map.getServiceOptions();
    }

    public static setServiceOptions(mapId: string, serviceOptions: atlas.ServiceOptions): void {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        map.setServiceOptions(serviceOptions);
    }
    // #endregion

    // #region Traffic
    public static getTraffic(mapId: string): atlas.TrafficOptions {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        return map.getTraffic();
    }

    public static setTraffic(mapId: string, options: atlas.TrafficOptions) {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        map.setTraffic(options);
    }
    // #endregion

    // #region Style
    public static getStyle(mapId: string): any {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        return map.getStyle();
    }

    public static setStyle(mapId: string, style: atlas.StyleOptions): void {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        map.setStyle(style);
    }
    // #endregion

    // #region User Interaction
    public static getUserInteraction(mapId: string): any {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }

        return map.getUserInteraction();
    }

    public static setUserInteraction(mapId: string, userInteraction: atlas.UserInteractionOptions): void {
        const map = AzFactory.getMap(mapId);
        if (!map) {
            return;
        }
        map.setUserInteraction(userInteraction);
    }
    // #endregion
}

type TCameraOptionsSet = (atlas.CameraOptions | (atlas.CameraBoundsOptions & { pitch?: number, bearing?: number })) & atlas.AnimationOptions;

interface CameraOptionsSet {
    camera?: atlas.CameraOptions;
    cameraBounds?: atlas.CameraBoundsOptions & { pitch?: number, bearing?: number };
    animation?: atlas.AnimationOptions;
}

interface MapOptionsSet {
    camera?: CameraOptionsSet;
    service?: atlas.ServiceOptions;
    style?: atlas.StyleOptions;
    userInteraction?: atlas.UserInteractionOptions;
}
